py4cast/models/vision/unetrpp.py

```diff
@@ -266,21 +291,38 @@ def forward(self, x):
         q_shared = torch.nn.functional.normalize(q_shared, dim=-1)
         k_shared = torch.nn.functional.normalize(k_shared, dim=-1)
 
-        attn_CA = (q_shared @ k_shared.transpose(-2, -1)) * self.temperature
-        attn_CA = attn_CA.softmax(dim=-1)
-        attn_CA = self.attn_drop(attn_CA)
-        x_CA = (attn_CA @ v_CA).permute(0, 3, 1, 2).reshape(B, N, C)
-
-        attn_SA = (
-            q_shared.permute(0, 1, 3, 2) @ k_shared_projected
-        ) * self.temperature2
-        attn_SA = attn_SA.softmax(dim=-1)
-        attn_SA = self.attn_drop_2(attn_SA)
-        x_SA = (
-            (attn_SA @ v_SA_projected.transpose(-2, -1))
-            .permute(0, 3, 1, 2)
-            .reshape(B, N, C)
-        )
+        if self.use_scaled_dot_product_CA:
+            x_CA = scaled_dot_product_attention(
+                q_shared, k_shared, v_CA, dropout_p=self.attn_drop.p
+            )
+        else:
+            attn_CA = (q_shared @ k_shared.transpose(-2, -1)) * self.temperature
+            attn_CA = attn_CA.softmax(dim=-1)
+            attn_CA = self.attn_drop(attn_CA)
+            x_CA = attn_CA @ v_CA
+
+        x_CA = x_CA.permute(0, 3, 1, 2).reshape(B, N, C)
+
+        if self.use_scaled_dot_product_SA:
+            q_shared_projected = torch.einsum("bhdn,nk->bhdk", q_shared, self.EF)
+            x_SA = scaled_dot_product_attention(
+                q_shared_projected.permute(0, 1, 3, 2),
+                k_shared_projected.permute(0, 1, 3, 2),
+                v_SA_projected.permute(0, 1, 3, 2),
+                dropout_p=self.attn_drop_2.p,
+            )
+            x_SA = torch.einsum("bhdn,nk->bhdk", x_SA.permute(0, 1, 3, 2), self.FE)
+
+        else:
+            attn_SA = (
+                q_shared.permute(0, 1, 3, 2) @ k_shared_projected
```

Comment:
why is `q_shared_projected` computed only in sdpa ?

Reply:
passing the matrices (q_shared, k_shared_projected, ...) used in the original implementation for spatial attention directly to sdpa leads to exceptions because the shape weren't right. So I decided to also project q_shared in the same subspace and permute to have a "spatial attention", "after projection" done by sdpa. If you have other ideas I am happy to try.

Reply:
Ok, so after discussion with PR owner, here's the pain:

-  sdpa requires same size q /k matrices, so, does not like k_projected and q_shared
- the current solution actually increases costs due to `q_shared_projected@k_shared_projected.transpose` which is O(D²) (with D the spatial dim). Any monkey-patch will result in the O(D²) popping somewhere.
- there might be a final shape mismatch in `x_SA` (not happening because of the channels/dims consistent evolution)

Reply:
Done. I launched a cheap experiment to compare results. thanks @flyIchtus